File: examples/cpp/example_07_camera.cpp
#include <array>
#include <glad/gl.h>

#include <renderer/window/window_t.hpp>
#include <renderer/assets/shader_manager_t.hpp>
#include <renderer/camera/camera_t.hpp>
#include <renderer/core/vertex_array_t.hpp>
#include <renderer/core/vertex_buffer_t.hpp>

#if defined(RENDERER_IMGUI)
#include <imgui.h>
#endif  // RENDERER_IMGUI

constexpr int WINDOW_WIDTH = 800;
constexpr int WINDOW_HEIGHT = 600;

auto main() -> int {
    renderer::Window window(WINDOW_WIDTH, WINDOW_HEIGHT);
    renderer::ShaderManager shader_manager;

    auto program = shader_manager.LoadProgram(
        "basic3d",
        renderer::EXAMPLES_PATH + "/resources/shaders/basic3d_vert.glsl",
        renderer::EXAMPLES_PATH + "/resources/shaders/basic3d_frag.glsl");

    // clang-format off
    std::array<float, 9> buffer_data = {
        /*|    pos     |*/
        -1.0F, 0.0F, -1.0F,
         1.0F, 0.0F, -1.0F,
         0.0F, 0.0F,  1.0F
    };
    // clang-format on

    renderer::BufferLayout layout = {
        {"position", renderer::eElementType::FLOAT_3, false}};

    auto vbo = std::make_unique<renderer::VertexBuffer>(
        layout, renderer::eBufferUsage::STATIC,
        static_cast<uint32_t>(sizeof(buffer_data)), buffer_data.data());

    auto vao = std::make_unique<renderer::VertexArray>();
    vao->AddVertexBuffer(std::move(vbo));

    renderer::Camera camera({0.0F, -3.0F, 0.0F}, {0.0F, 0.0F, 0.0F});
    Vec3 color = {0.1F, 0.2F, 0.7F};

    while (window.active()) {
        window.Begin();
        program->Bind();
        vao->Bind();

        program->SetMat4("u_model_matrix", Mat4::Identity());
        program->SetMat4("u_view_matrix", camera.view_matrix());
        program->SetMat4("u_proj_matrix", camera.proj_matrix());

#if defined(RENDERER_IMGUI)
        ImGui::Begin("Options");
        {
            // Allow the user to set the color of the triangle :D
            ImGui::ColorEdit3("color", color.data());
            // Allow the user to control the position of the camera
            static Vec3 s_camera_position = camera.position();
            ImGui::SliderFloat3("camera.position", s_camera_position.data(),
                                -10.0F, 10.0F);
            camera.SetPosition(s_camera_position);
        }
#endif  // RENDERER_IMGUI
        program->SetVec3("u_color", color);

        glDrawArrays(GL_TRIANGLES, 0, 3);

        vao->Unbind();
        program->Unbind();
        window.End();
    }
}

File: examples/python/example_07_camera.py
import os

import numpy as np
from OpenGL import GL

import renderer as rdr

WINDOW_WIDTH = 800
WINDOW_HEIGHT = 600

CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
SHADERS_DIR = os.path.join(CURRENT_DIR, "..", "resources", "shaders")


def main() -> int:
    window = rdr.Window(
        WINDOW_WIDTH, WINDOW_HEIGHT, rdr.WindowBackend.TYPE_GLFW
    )

    def keyboard_callback(key, action, mods) -> None:
        if key == rdr.Keys.KEY_ESCAPE and action == rdr.KeyAction.RELEASED:
            window.RequestClose()

    window.RegisterKeyboardCallback(keyboard_callback)

    shader_manager = rdr.ShaderManager()

    program = shader_manager.LoadProgram(
        "basic3d",
        os.path.join(SHADERS_DIR, "basic3d_vert.glsl"),
        os.path.join(SHADERS_DIR, "basic3d_frag.glsl"),
    )

    # fmt: off
    buffer_data = np.array([
        # |----pos----|
        -1.0, 0.0, -1.0,  # noqa: E131
         1.0, 0.0, -1.0,  # noqa: E131
         0.0, 0.0,  1.0,  # noqa: E131
    ], dtype=np.float32)
    # fmt: on

    layout = rdr.BufferLayout([["position", rdr.ElementType.FLOAT_3, False]])

    vbo = rdr.VertexBuffer(layout, rdr.BufferUsage.STATIC, buffer_data)

    vao = rdr.VertexArray()
    vao.AddVertexBuffer(vbo)

    camera = rdr.Camera([0.0, -3.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 1.0])
    color = np.array([0.1, 0.2, 0.7], dtype=np.float32)

    while window.active:
        window.Begin()
        program.Bind()
        vao.Bind()

        program.SetMat4("u_model_matrix", np.identity(4, dtype=np.float32))
        program.SetMat4("u_view_matrix", camera.view_matrix())
        program.SetMat4("u_proj_matrix", camera.proj_matrix())
        program.SetVec3("u_color", color)

        GL.glDrawArrays(GL.GL_TRIANGLES, 0, 3)

        vao.Unbind()
        program.Unbind()
        window.End()

    return 0


if __name__ == "__main__":
    SystemExit(main())
